cmd/soroban-rpc/internal/db/event.go

```diff
@@ -97,3 +164,185 @@ func (eventHandler *eventHandler) InsertEvents(lcm xdr.LedgerCloseMeta) error {
 
 	return nil
 }
+
+type ScanFunction func(
+	event xdr.DiagnosticEvent,
+	cursor Cursor,
+	ledgerCloseTimestamp int64,
+	txHash *xdr.Hash,
+) bool
+
+// trimEvents removes all Events which fall outside the ledger retention window.
+func (eventHandler *eventHandler) trimEvents(latestLedgerSeq uint32, retentionWindow uint32) error {
+	if latestLedgerSeq+1 <= retentionWindow {
+		return nil
+	}
+	cutoff := latestLedgerSeq + 1 - retentionWindow
+	id := Cursor{Ledger: cutoff}.String()
+
+	_, err := sq.StatementBuilder.
+		RunWith(eventHandler.stmtCache).
+		Delete(eventTableName).
+		Where(sq.Lt{"id": id}).
+		Exec()
+	return err
+}
+
+// GetEvents applies f on all the events occurring in the given range with specified contract IDs if provided.
+// The events are returned in sorted ascending Cursor order.
+// If f returns false, the scan terminates early (f will not be applied on
+// remaining events in the range).
+//
+//nolint:funlen
+func (eventHandler *eventHandler) GetEvents(
+	ctx context.Context,
+	cursorRange CursorRange,
+	contractIDs [][]byte,
+	topics [][]string,
+	eventTypes []int,
+	f ScanFunction,
+) error {
+	start := time.Now()
+
+	rowQ := sq.
+		Select(" id", "event_data", "transaction_hash", "ledger_close_time").
+		From(eventTableName).
+		Where(sq.GtOrEq{"id": cursorRange.Start.String()}).
+		Where(sq.Lt{"id": cursorRange.End.String()}).
+		OrderBy("id ASC")
+
+	if len(contractIDs) > 0 {
+		rowQ = rowQ.Where(sq.Eq{"contract_id": contractIDs})
+	}
+	if len(eventTypes) > 0 {
+		rowQ = rowQ.Where(sq.Eq{"event_type": eventTypes})
+	}
+
+	if len(topics) > 0 {
+		var orConditions sq.Or
+		for i, topic := range topics {
+			if topic == nil {
+				continue
+			}
+			orConditions = append(orConditions, sq.Eq{fmt.Sprintf("topic%d", i+1): topic})
```

Comment:
How does this work for wildcard topic filters?

Reply:
![image](https://github.com/user-attachments/assets/1c56be50-87a7-4847-ab0c-4f1c650b9ec6)

Attaching small design diagram to explain it


1. Combine Filters: We ignore the wildcard topic entry when we combine all the filters together. This might be the only way because what we need to tell DB is to fetch all the entries when one of the topic is wildcard

2. Apply filter after fetching events from DB: In this step we will check the actual intersection of topics e.g where topic1 = "some bytes" and topic2 = " wildcard(*) " and topic3 = "some bytes"
note that events that we fetch from DB always does not end up in result set as we need to check for the final intersection of all the filters for given EventXDR.

I went with this approach in order to reduce latency and multiple DB calls. @Shaptic

Reply:
Got it! And we're doing the second-stage filtering client side to ensure the topics are OR'd not AND'd? Per the [docs](https://developers.stellar.org/docs/data/rpc/api-reference/methods/getEvents) (had to double-check myself):

> Events matching **any** of the filters are included. To match a filter, an event must match both a contractId and a topic.

Reply:
Yes second-stage filtering on service side. Well client side if you think in the context of DB hahaa 

Code:

```
if request.Matches(event) { --> This is 2nd stage filtering to exactly match the filter
			found = append(found, entry{cursor, ledgerCloseTimestamp, event, txHash})
		}

```

https://github.com/stellar/soroban-rpc/pull/215/files#diff-1f027a1543c912ae4bc1533d9fece665a2b555548e8f74a8e591a70fdc317b2fR475

Reply:
And this is the interesting part so wanted to make it as clear as possible.

Reply:
Nice, I didn't see that line! Huge diff :laughing:

Reply:
hahaa yes sorry for the big diff 😆